drivers/sdhc/brcm_bcm2712_sdhci.c

```diff
@@ -0,0 +1,484 @@
+		LOG_DEV_DBG(dev, "sd detect gpio %d", cd);
+		goto out_cd;
+	}
+
+	cd = sdhci_get_cd(&data->sdhci_ctx);
+
+out_cd:
+	LOG_DEV_DBG(dev, "sd detect sdhci %d", cd);
+
+	if (!cd) {
+		LOG_DEV_INF(dev, "sd card is not detected");
+	}
+	return cd;
+}
+
+static int bcm2712_sdhci_card_busy(const struct device *dev)
+{
+	struct bcm2712_sdhci_data *data = dev->data;
+
+	return sdhci_is_card_busy(&data->sdhci_ctx);
+}
+
+static int bcm2712_sdhci_set_io(const struct device *dev, struct sdhc_io *ios)
+{
+	const struct bcm2712_sdhci_config *cfg = dev->config;
+	struct bcm2712_sdhci_data *data = dev->data;
+	struct sdhc_io *host_io = &data->host_io;
+	int ret;
+
+	LOG_DEV_DBG(dev,
+		    "set_io: bus_width:%d clock:%dHz power:%d timing:%d, voltage:%d drv:%d",
+		    ios->bus_width, ios->clock, ios->power_mode,
+		    ios->timing, ios->signal_voltage, ios->driver_type);
+
+	if (host_io->clock != ios->clock) {
+
+		if (ios->clock &&
+		    (ios->clock > data->props.f_max || ios->clock < data->props.f_min)) {
+			LOG_DEV_ERR(dev, "set_io: invalid clock:%d valid range %d - %d Hz",
+				    ios->clock, data->props.f_min, data->props.f_max);
+			return -EINVAL;
+		}
+
+		ret = sdhci_set_clock(&data->sdhci_ctx, ios->clock);
+		if (ret) {
+			LOG_DEV_ERR(dev, "set_io: set clock failed (%d)", ret);
+			return ret;
+		}
+		host_io->clock = ios->clock;
+	}
+
+	if (host_io->bus_width != ios->bus_width) {
+		ret = sdhci_set_bus_width(&data->sdhci_ctx, ios->bus_width);
+		if (ret) {
+			LOG_DEV_ERR(dev, "set_io: set bus width failed (%d)", ret);
+			return ret;
+		}
+		host_io->bus_width = ios->bus_width;
+	}
+
+	if (host_io->power_mode != ios->power_mode) {
+		if (ios->power_mode == SDHC_POWER_ON) {
+			regulator_enable(cfg->regulator_vmmc);
+
+			k_msleep(cfg->power_delay_ms);
+
+			sdhci_set_bus_power(&data->sdhci_ctx, SD_VOL_3_3_V);
+
+			regulator_enable(cfg->regulator_vqmmc);
+		} else if (ios->power_mode == SDHC_POWER_OFF) {
+			sdhci_set_bus_power(&data->sdhci_ctx, 0);
+
+			if (regulator_is_enabled(cfg->regulator_vqmmc)) {
+				regulator_disable(cfg->regulator_vqmmc);
+			}
+			if (regulator_is_enabled(cfg->regulator_vmmc)) {
+				regulator_disable(cfg->regulator_vmmc);
+			}
+		} else {
+			return -EINVAL;
+		}
+		host_io->power_mode = ios->power_mode;
+	}
+
+	if (host_io->signal_voltage != ios->signal_voltage) {
+		if (ios->signal_voltage == SD_VOL_3_3_V) {
+			ret = regulator_set_voltage(cfg->regulator_vqmmc, 3300000, 3300000);
+		} else if (ios->signal_voltage == SD_VOL_1_8_V) {
+			ret = regulator_set_voltage(cfg->regulator_vqmmc, 1800000, 1800000);
+		} else {
+			return -EINVAL;
+		}
+
+		ret = sdhci_set_voltage(&data->sdhci_ctx, ios->signal_voltage);
+		if (ret) {
+			LOG_DEV_ERR(dev, "set_io: set signal voltage failed (%d)", ret);
+			return ret;
+		}
+		host_io->signal_voltage = ios->signal_voltage;
+	}
+
+	if (host_io->timing != ios->timing) {
+		ret = sdhci_set_uhs_timing(&data->sdhci_ctx, ios->timing);
```

Comment:
The  SDHC parameter name "UHS Mode Select" the input parameter name "timing". Not sure what's strange. Also  it follows uboot/linux notation.

Reply:
Reading the name of this common function, I, as a reviewer of the code, thought that it is designed for UHS timing modes only because the name suggests that. But what about High Speed and Normal timing modes?

For instance, n the SDHCI init function, we have the following line:
```
host_caps->high_spd_support = sdhci_ctx->caps & SDHCI_CAN_DO_HISPD ? 1 : 0;
```

Okay, I see that you use `sdhci_set_uhs_timing` only for setting UHS modes, but according to the SD Host Controller Simplified Specification version 3.0 and even according to your SDHCI init code, it is possible to have non-UHS timing modes (look at section 3.9 'Changing Bus Speed Mode' from the aforementioned specification).

Reply:
ok. I created TODO "sdhc: rename sdhci_set_uhs_timing to sdhci_set_timing"

Reply:
or you can create an issue for our Zephyr repo with link to this comment

Reply:
https://github.com/orgs/xen-troops/projects/6/views/1?pane=issue&itemId=69912481